### PR TITLE
ddl: args v2 for create/alter/drop placement policy 

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: ref #53930

Problem Summary:

### What changed and how does it work?

As title.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/executor.go [hunk 1]
```diff
@@ -1357,6 +1357,7 @@ func (e *executor) CreatePlacementPolicyWithInfo(ctx sessionctx.Context, policy
 	policy.ID = policyID
 
 	job := &model.Job{
+		Version:    model.GetJobVerInUse(),
 		SchemaName: policy.Name.L,
 		Type:       model.ActionCreatePlacementPolicy,
 		BinlogInfo: &model.HistoryInfo{},
```
Comment (joccau) [inline pkg/ddl/executor.go:1363]: need remove `Args: []any{policy, onExist == OnExistReplace},`

File: pkg/ddl/placement_policy.go [hunk 2]
```diff
@@ -183,7 +184,11 @@ func checkAllTablePlacementPoliciesExistAndCancelNonExistJob(t *meta.Meta, job *
 }
 
 func onDropPlacementPolicy(jobCtx *jobContext, t *meta.Meta, job *model.Job) (ver int64, _ error) {
-	policyInfo, err := checkPlacementPolicyExistAndCancelNonExistJob(t, job, job.SchemaID)
+	args, err := model.GetPlacementPolicyArgs(job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	policyInfo, err := checkPlacementPolicyExistAndCancelNonExistJob(t, job, args.PolicyID)
```
Comment (joccau) [inline pkg/ddl/placement_policy.go:191]: Why need to use args.PolicyID replace job.SchemaID ?

File: pkg/ddl/placement_policy_ddl_test.go [hunk 3]
```diff
@@ -44,13 +44,18 @@ func testPlacementPolicyInfo(t *testing.T, store kv.Storage, name string, settin
 
 func testCreatePlacementPolicy(t *testing.T, ctx sessionctx.Context, d ddl.ExecutorForTest, policyInfo *model.PolicyInfo) *model.Job {
 	job := &model.Job{
+		Version:    model.GetJobVerInUse(),
 		SchemaName: policyInfo.Name.L,
 		Type:       model.ActionCreatePlacementPolicy,
 		BinlogInfo: &model.HistoryInfo{},
 		Args:       []any{policyInfo},
```
Comment (joccau) [inline pkg/ddl/placement_policy_ddl_test.go:51]:
```suggestion
```

File: pkg/meta/model/job_args.go [hunk 4]
```diff
@@ -779,3 +779,51 @@ func GetAddCheckConstraintArgs(job *Job) (*AddCheckConstraintArgs, error) {
 	}
 	return getOrDecodeArgsV2[*AddCheckConstraintArgs](job)
 }
+
+// PlacementPolicyArgs is the argument for create/alter/drop placement policy
+type PlacementPolicyArgs struct {
+	Policy         *PolicyInfo  `json:"policy,omitempty"`
+	ReplaceOnExist bool         `json:"replace_on_exist,omitempty"`
+	PolicyName     pmodel.CIStr `json:"policy_name,omitempty"`
+	PolicyID       int64        `json:"policy_id"`
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:788]:
```suggestion
	// it's set for alter/drop policy in v2
	PolicyID       int64        `json:"policy_id"`
```

File: pkg/meta/model/job_args.go [hunk 5]
```diff
@@ -779,3 +779,51 @@ func GetAddCheckConstraintArgs(job *Job) (*AddCheckConstraintArgs, error) {
...
+		}
+		return
+	}
+	job.Args = []any{a}
+}
+
+// GetRenameTablesArgs gets the placement policy args.
+func GetPlacementPolicyArgs(job *Job) (*PlacementPolicyArgs, error) {
+	if job.Version == JobVersion1 {
+		var args *PlacementPolicyArgs = &PlacementPolicyArgs{PolicyID: job.SchemaID}
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:809]:
```suggestion
		args := &PlacementPolicyArgs{PolicyID: job.SchemaID}
```

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/ddl/placement_policy.go (hunk 2 above)
Comment (joechenrh) [reply pkg/ddl/placement_policy.go:191]: Just want to put all related variables into arguments. I think using `job.SchemaID` to store `PolicyID` is a bit confusing.

**Review by D3Hunter — APPROVED** [review]

**Comment by D3Hunter** [comment]
there's panic
```
[2024/09/23 05:48:40.041 +00:00] [ERROR] [misc.go:115] ["panic in the recoverable goroutine"] [label=ddl-worker] [funcInfo="worker 17, tp general runOneJobStep"] [r="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.Recover\n\tpkg/util/misc.go:119\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.panicmem\n\tGOROOT/src/runtime/panic.go:261\nruntime.sigpanic\n\tGOROOT/src/runtime/signal_unix.go:861\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onDropTablePartition\n\tpkg/ddl/partition.go:2191\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onAddTablePartition\n\tpkg/ddl/partition.go:100\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).runOneJobStep\n\tpkg/ddl/job_worker.go:867\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).transitOneJobStep\n\tpkg/ddl/job_worker.go:559\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).transitOneJobStepAndWaitSync\n\tpkg/ddl/job_scheduler.go:577\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).deliveryJob.func1\n\tpkg/ddl/job_scheduler.go:492\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1\n\tpkg/util/wait_group_wrapper.go:157"] [stack="github.com/pingcap/tidb/pkg/util.Recover\n\tpkg/util/misc.go:115\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.panicmem\n\tGOROOT/src/runtime/panic.go:261\nruntime.sigpanic\n\tGOROOT/src/runtime/signal_unix.go:861\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onDropTablePartition\n\tpkg/ddl/partition.go:2191\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onAddTablePartition\n\tpkg/ddl/partition.go:100\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).runOneJobStep\n\tpkg/ddl/job_worker.go:867\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).transitOneJobStep\n\tpkg/ddl/job_worker.go:559\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).transitOneJobStepAndWaitSync\n\tpkg/ddl/job_scheduler.go:577\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).deliveryJob.func1\n\tpkg/ddl/job_scheduler.go:492\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1\n\tpkg/util/wait_group_wrapper.go:157"]
```

**Comment by joechenrh** [comment]
> there's panic
> 
> ```
> [2024/09/23 05:48:40.041 +00:00] [ERROR] [misc.go:115] ["panic in the recoverable goroutine"] [label=ddl-worker] [funcInfo="worker 17, tp general runOneJobStep"] [r="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.Recover\n\tpkg/util/misc.go:119\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.panicmem\n\tGOROOT/src/runtime/panic.go:261\nruntime.sigpanic\n\tGOROOT/src/runtime/signal_unix.go:861\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onDropTablePartition\n\tpkg/ddl/partition.go:2191\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onAddTablePartition\n\tpkg/ddl/partition.go:100\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).runOneJobStep\n\tpkg/ddl/job_worker.go:867\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).transitOneJobStep\n\tpkg/ddl/job_worker.go:559\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).transitOneJobStepAndWaitSync\n\tpkg/ddl/job_scheduler.go:577\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).deliveryJob.func1\n\tpkg/ddl/job_scheduler.go:492\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1\n\tpkg/util/wait_group_wrapper.go:157"] [stack="github.com/pingcap/tidb/pkg/util.Recover\n\tpkg/util/misc.go:115\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.panicmem\n\tGOROOT/src/runtime/panic.go:261\nruntime.sigpanic\n\tGOROOT/src/runtime/signal_unix.go:861\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onDropTablePartition\n\tpkg/ddl/partition.go:2191\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).onAddTablePartition\n\tpkg/ddl/partition.go:100\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).runOneJobStep\n\tpkg/ddl/job_worker.go:867\ngithub.com/pingcap/tidb/pkg/ddl.(*worker).transitOneJobStep\n\tpkg/ddl/job_worker.go:559\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).transitOneJobStepAndWaitSync\n\tpkg/ddl/job_scheduler.go:577\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).deliveryJob.func1\n\tpkg/ddl/job_scheduler.go:492\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1\n\tpkg/util/wait_group_wrapper.go:157"]
> ```

This is because `PartInfo` is not set when rollback?

https://github.com/pingcap/tidb/blob/d55808c15e99c1374365a589549863fb906a4e47/pkg/ddl/partition.go#L2159-L2165

https://github.com/pingcap/tidb/blob/d55808c15e99c1374365a589549863fb906a4e47/pkg/meta/model/job_args.go#L454-L466

**Comment by D3Hunter** [comment]
panic will be fixed here https://github.com/pingcap/tidb/pull/56222

**Review by joccau — APPROVED** [review]
LGTM

**Pull request merged**